### PR TITLE
Challenge: Kubernetes and documentation

### PR DESCRIPTION
**Pull request by rihp**
### PR Quality Checklist
- [ ] My pull request is atomic and focuses on a single change.
- [ ] I have thoroughly tested my changes with multiple different prompts.
- [ ] I have considered potential risks and mitigations for my changes.
- [ ] I have documented my changes clearly and comprehensively.
- [ ] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

File: docs/challenges/building_challenges.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+Join us in shaping the future of Auto-GPT by designing challenges that test its limits. Your input will be invaluable in guiding our progress and ensuring that we're on the right track. We're seeking individuals with a diverse skill set, including:
+
+🎨 UX Design: Your expertise will enhance the user experience for those attempting to conquer our challenges. With your help, we'll develop a dedicated section in our wiki, and potentially even launch a standalone website.
+
+💻 Coding Skills: Proficiency in Python, pytest, and VCR (a library that records OpenAI calls and stores them) will be essential for creating engaging and robust challenges.
+
+⚙️ DevOps Skills: Experience with CI pipelines in GitHub and possibly Google Cloud Platform will be instrumental in streamlining our operations.
+
+Are you ready to play a pivotal role in Auto-GPT's journey? Apply now to become a Challenge Creator! 🚀
```
Comment (samuelbutler) [inline docs/challenges/building_challenges.md]: Why do people need to apply to be a challenge creator? Why wouldn't we structure this so anyone can submit a challenge?

File: docs/challenges/building_challenges.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+## Defining your Agent
+
+Go to https://github.com/Significant-Gravitas/Auto-GPT/blob/master/tests/integration/agent_factory.py
+
+Create your agent fixture, in line 8 through 13 we define all the settings of our challenged agent. We also give it a name in line 20.
+
+Please choose the commands your agent will need to beat the challenges, the full list is available in the main.py See lines 4 to 6. (we 're working on a better way to design this, for now you have to look at main.py)
+
+```python=
```
Comment (samuelbutler) [inline docs/challenges/building_challenges.md]: I think this stray `=` is breaking the syntax highlighting for this code block

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    input_sequence = ["s", "s", "s", "s", "s", "EXIT"]
+    gen = input_generator(input_sequence)
+    monkeypatch.setattr("builtins.input", lambda _: next(gen))
+
+    with contextlib.suppress(SystemExit):
+        run_interaction_loop(kubernetes_agent, None)
+
+    file_path = str(kubernetes_agent.workspace.get_path("kube.yaml"))
+    content = read_file(file_path)
+    with open('kube.yaml') as file:
```
Comment (samuelbutler) [inline tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: Won't `open('kube.yaml') `only work if this file is being run within the auto_gpt_workspace?

File: docs/challenges/building_challenges.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        try:
+            yaml.safe_load(file)
+        except yaml.YAMLError as e:
+            assert False, f"Error while loading kube.yaml: {e}"
+    assert "nginx" in content, "Expected the file to contain nginx"
+
+
+```
+
+Check how in lines 45 to 52 we try to check if the file is a proper yaml
```
Comment (samuelbutler) [inline docs/challenges/building_challenges.md]: I assume this is a note to reviewers? If not then those line numbers will be meaningless to readers bc this markdown document won't have line numbers in the code blocks.

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import contextlib
+from functools import wraps
```
Comment (samuelbutler) [inline tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: linter complaints: `functools.wraps` imported but unused

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+from typing import Generator
+
+import pytest
+import yaml
+
+from autogpt.commands.file_operations import read_file, write_to_file
```
Comment (samuelbutler) [inline tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: linter complaint: autogpt.commands.file_operations.write_to_file` imported but unused

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+# @pytest.skip("Nobody beat this challenge yet")
+# @pytest.mark.skip("This challenge hasn't been beaten yet.")
+@pytest.mark.vcr
+@requires_api_key("OPENAI_API_KEY")
+@run_multiple_times(3)
+def test_information_retrieval_challenge_a(
+    kubernetes_agent, monkeypatch
+) -> None:
+    """
+    Test the challenge_a function in a given agent by mocking user inputs and checking the output file content.
```
Comment (samuelbutler) [inline tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: linter complaint: line too long

File: docs/challenges/building_challenges.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    return agent
+```
+
+## Creating your challenge
+Go to `tests/integration/challenges`and create a file that is called `test_your_test_description.py` and add it to the appropriate folder. If no category exists you can create a new one.
+
+Your test could look something like this 
+
+```python=
```
Comment (samuelbutler) [inline docs/challenges/building_challenges.md]: same stray `=` is breaking the syntax highlighting

**Review by samuelbutler — COMMENTED** [review]
Hey Media, 

I'm trying to get some experience doing code reviews, and this seemed like a good opportunity. So, don't put too much weight behind any of my comments. 

Let me know if there is anything I could do to clarify my comments.

**Review by samuelbutler — COMMENTED** [review]
Hey Media, 

I'm trying to get some experience doing code reviews, and this seemed like a good opportunity. So, don't put too much weight behind any of my comments. 

Let me know if there is anything I could do to clarify my comments.

File: docs/challenges/building_challenges.md (hunk 1 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: I copied this from the wiki, but i agree this should always be open

File: docs/challenges/building_challenges.md (hunk 2 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: Weird, i will remove it. It should add a line counter to the codeblock. thanks

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py (hunk 3 above)
Comment (rihp) [reply tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: You're right, i cant get this test to actually load up the file from the workspace currently

looking for a way to resolve this

File: docs/challenges/building_challenges.md (hunk 4 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: yes, if this is the case i will change the comment to work without the line counters

File: docs/challenges/building_challenges.md (hunk 8 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: removed

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py (hunk 5 above)
Comment (rihp) [reply tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: removed

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py (hunk 7 above)
Comment (rihp) [reply tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: updated

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py (hunk 6 above)
Comment (rihp) [reply tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: removed

File: docs/challenges/building_challenges.md (hunk 4 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: updated

File: docs/challenges/building_challenges.md (hunk 1 above)
Comment (waynehamadi) [reply docs/challenges/building_challenges.md]: yeah that's fine, "applications" are informal, it's just people creating PRs.

File: tests/integration/agent_factory.py [hunk 9]
```diff
@@ -183,3 +183,69 @@ def get_company_revenue_agent(
...
+        system_prompt=system_prompt,
+        triggering_prompt=DEFAULT_TRIGGERING_PROMPT,
+        workspace_directory=workspace.root,
+    )
+
+    return agent
+
+
+@pytest.fixture
+def translation_agent(agent_test_config, memory_local_cache, workspace: Workspace):
```
Comment (waynehamadi) [inline tests/integration/agent_factory.py]: remove that one you're not using it

File: docs/challenges/building_challenges.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    Creates a generator that yields input strings from the given sequence.
+
+    :param input_sequence: A list of input strings.
+    :return: A generator that yields input strings.
+    """
+    yield from input_sequence
+
+
+# @pytest.skip("Nobody beat this challenge yet")
+# @pytest.mark.skip("This challenge hasn't been beaten yet.")
```
Comment (waynehamadi) [inline docs/challenges/building_challenges.md]: skip it please

**Comment by merwanehamadi** [comment]
@rihp great thanks!
The success criteria could probably be improved, but I don't know how yet, let's merge once you resolve comments

**Review by merwanehamadi — CHANGES_REQUESTED** [review]

File: docs/challenges/building_challenges.md (hunk 10 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: skipped

File: tests/integration/agent_factory.py (hunk 9 above)
Comment (rihp) [reply tests/integration/agent_factory.py]: done

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py (hunk 3 above)
Comment (rihp) [reply tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py]: changed the implementation

File: docs/challenges/building_challenges.md (hunk 1 above)
Comment (rihp) [reply docs/challenges/building_challenges.md]: updated

File: tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    Creates a generator that yields input strings from the given sequence.
+
+    :param input_sequence: A list of input strings.
+    :return: A generator that yields input strings.
+    """
+    yield from input_sequence
+
+
+@pytest.skip("Nobody beat this challenge yet")
+@pytest.mark.skip("This challenge hasn't been beaten yet.")
```
Comment (waynehamadi) [inline tests/integration/challenges/kubernetes/test_kubernetes_template_challenge_a.py:23]: remove @pytest.skip("Nobody beat this challenge yet")

File: docs/challenges/building_challenges.md [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    command_registry = CommandRegistry()
+    command_registry.import_commands("autogpt.commands.file_operations")
+    command_registry.import_commands("autogpt.app")
+
+    # Define all the settings of our challenged agent
+    ai_config = AIConfig(
+        ai_name="Kubernetes",
+        ai_role="an autonomous agent that specializes in creating Kubernetes deployment templates.",
+        ai_goals=[
+            "Write the a simple kubernetes deployment file and save it as kube.yaml. You should make a simple nginx web server that uses docker and exposes the port 80.",
```
Comment (waynehamadi) [inline docs/challenges/building_challenges.md]: "the a"
"as a kube.yaml"

File: tests/integration/agent_factory.py [hunk 13]
```diff
@@ -183,3 +183,36 @@ def get_company_revenue_agent(
...
+def kubernetes_agent(memory_local_cache, workspace: Workspace):
+    command_registry = CommandRegistry()
+    command_registry.import_commands("autogpt.commands.file_operations")
+    command_registry.import_commands("autogpt.app")
+
+    ai_config = AIConfig(
+        ai_name="Kubernetes",
+        ai_role="an autonomous agent that specializes in creating Kubernetes deployment templates.",
+        ai_goals=[
+            "Write the a simple kubernetes deployment file and save it as kube.yaml.",
```
Comment (waynehamadi) [inline tests/integration/agent_factory.py]: the a

as a

**Review by merwanehamadi — CHANGES_REQUESTED** [review]

**Review by merwanehamadi — APPROVED** [review]

**Pull request merged**